grype/vulnerability_matcher.go

```diff
@@ -314,37 +314,42 @@ func (m *VulnerabilityMatcher) normalizeByCVE(match match.Match) match.Match {
 // ignoreRulesByLocation implements match.IgnoreFilter to filter each matching
 // package that overlaps by location and have the same vulnerability ID (CVE)
 type ignoreRulesByLocation struct {
+	remainingFilters      []match.IgnoreFilter
 	locationToIgnoreRules map[string][]match.IgnoreRule
 }
 
 func (i ignoreRulesByLocation) IgnoreMatch(m match.Match) []match.IgnoreRule {
 	for _, l := range m.Package.Locations.ToSlice() {
 		for _, rule := range i.locationToIgnoreRules[l.RealPath] {
-			if rule.Vulnerability == m.Vulnerability.ID {
-				return []match.IgnoreRule{rule}
-			}
-			for _, relatedVulnerability := range m.Vulnerability.RelatedVulnerabilities {
-				if rule.Vulnerability == relatedVulnerability.ID {
-					return []match.IgnoreRule{rule}
-				}
+			if matched := rule.IgnoreMatch(m); matched != nil {
+				return matched
 			}
 		}
 	}
+	for _, f := range i.remainingFilters {
+		if matched := f.IgnoreMatch(m); matched != nil {
+			return matched
+		}
+	}
 	return nil
 }
 
-// ignoreMatchFilter creates an ignore filter based on the provided IgnoredMatches to filter out "the same"
+// ignoredMatchFilter creates an ignore filter based on the provided IgnoredMatches to filter out "the same"
 // vulnerabilities reported by other matchers based on overlapping file locations
-func ignoredMatchFilter(ignores []match.IgnoredMatch) match.IgnoreFilter {
+func ignoredMatchFilter(ignores []match.IgnoreFilter) match.IgnoreFilter {
 	out := ignoreRulesByLocation{locationToIgnoreRules: map[string][]match.IgnoreRule{}}
-	for _, ignore := range ignores {
-		// TODO should this be syftPkg.FileOwner interface or similar?
-		if m, ok := ignore.Package.Metadata.(pkg.ApkMetadata); ok {
-			for _, f := range m.Files {
-				out.locationToIgnoreRules[f.Path] = append(out.locationToIgnoreRules[f.Path], ignore.AppliedIgnoreRules...)
-			}
+	for i := 0; i < len(ignores); i++ {
+		rule, ok := ignores[i].(match.IgnoreRule)
+		if !ok {
+			continue
+		}
+		if rule.Package.Location != "" && !strings.ContainsRune(rule.Package.Location, '*') {
+			ignores = append(ignores[:i], ignores[i+1:]...) // this rule is handled with location lookups
+			i--
+			out.locationToIgnoreRules[rule.Package.Location] = append(out.locationToIgnoreRules[rule.Package.Location], rule)
 		}
 	}
+	out.remainingFilters = ignores // the remaining rules are not location-based rules
```

Comment:
nit, consider using slices.DeleteFunc or slices.Retain (new in go1.21) which is easier to read & harder to mess up with subtle index / append issues.